k2/python/k2/symbol_table.py

```diff
@@ -90,7 +106,55 @@ def from_file(filename: str) -> 'SymbolTable':
         with open(filename, 'r') as f:
             return SymbolTable.from_str(f.read().strip())
 
-    def get(self, k: Union[int, str]) -> Union[str, int]:
+    def to_file(self, filename: str):
+        '''Serialize the SymbolTable to a file.
+
+        Every line in the symbol table file has two fields separated by
+        space(s), tab(s) or both. The following is an example file:
+
+        .. code-block::
+
+            <eps> 0
+            a 1
+            b 2
+            c 3
+
+        Args:
+          filename:
+            Name of the symbol table file. Its format is documented above.
+        '''
+        with open(filename, 'w') as f:
+            for idx, symbol in sorted(self._id2sym.items()):
+                print(symbol, idx, file=f)
+
+    def add(self, symbol: Symbol, index: Optional[int] = None) -> int:
+        '''Add a new symbol to the SymbolTable.
+
+        Args:
+            symbol:
+                The symbol to be added.
+            index:
+                Optional int id to which the symbol should be assigned.
+                If it is not available, a ValueError will be raised.
+
+        Returns:
+            The int id to which the symbol has been assigned.
+        '''
+        # Already in the table? Return it's ID.
+        if symbol in self._sym2id:
+            return self._sym2id[symbol]
+        # Specific ID not provided - use next available.
+        if index is None:
+            index = len(self)
```

Comment:
It can happen that IDs are not contiguous and the ID returned by `len(self)`
may already be occupied.

Shall we add a field `next_available_id_` that is initialized to the largest ID + 1 ?